File: 25.reverse-nodes-in-k-group.py
#
# @lc app=leetcode id=25 lang=python3
#
# [25] Reverse Nodes in k-Group
#

# @lc code=start
# Definition for singly-linked list.
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        #Constraints:
        # 1 <= k <= n <= 5000
        # 0 <= Node.val <= 1000
        
        n = 0
        cur = head
        while cur:
            n += 1
            cur = cur.next
        
        dummy = pointer = ListNode(next = head)
        prev = None
        cur = head

        while n >= k:
            n -= k
            for _ in range(k):
                next_node = cur.next
                cur.next = prev
                prev = cur
                cur = next_node
            
            next_node = pointer.next
            next_node.next = cur
            pointer.next = prev
            pointer = next_node
        return dummy.next              
# @lc code=end

File: 25.reverse-nodes-in-k-group.java
/*
 * @lc app=leetcode id=25 lang=java
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        // int n = 0;
        // ListNode cnt = head;
        // ListNode dummy = new ListNode(0, head);
        // ListNode dummyhead = dummy;
        // ListNode cur = head;
        // while(cnt != null){
        //     n++;
        //     cnt = cnt.next;
        // }
        // int groups = n / k; // how many groups with length k we are able to reverse
        // while(groups > 0){
        //     ListNode prev = null;
        //     for(int i = 0; i < k; i++){
        //         ListNode next = cur.next;
        //         cur.next = prev;
        //         prev = cur;
        //         cur = next;
        //     }
        //     ListNode first = dummyhead.next;
        //     first.next = cur;
        //     dummyhead.next = prev;
        //     dummyhead = first;
        //     groups--; 
        // }
        // return dummy.next;

        int len = 0;
        ListNode cnt = head;
        while(cnt != null){
            len++;
            cnt = cnt.next;
        }
        int iter = len/k;

        ListNode dummy = new ListNode(0, head);
        ListNode tmp = dummy;
        ListNode cur = head;

        while(iter-- > 0){
            ListNode prev = null;
            for(int i = 0; i < k; i++){
                ListNode next = cur.next;
                cur.next = prev;
                prev = cur;
                cur = next;
            }
            ListNode tail = tmp.next;
            tail.next = cur;
            tmp.next = prev;
            tmp = tail;
        }
        return dummy.next;

        
        
    }
}
// @lc code=end

